cpp/src/arrow/compute/exec/exec_node_runner_impl.h

```diff
@@ -0,0 +1,63 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include "arrow/compute/exec/util.h"
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/util/future.h"
+
+namespace arrow {
+namespace compute {
+
+class ExecNodeRunnerImpl {
```

Comment:
Looking at the implementations, I wonder if this shouldn't actually just implement ExecNode? The reason being that most of the boilerplate we want to get rid of is still there right now, and additionally, there's more opportunity for implementer mistakes since you have to remember to call these new methods on the impl, and since the impl class hides the implementation, it's hard to tell at a glance if things are correct anymore. A subclass approach may not compose as well, but I'm not sure there's composition to be had in the first place here, and it would actually remove most of the boilerplate.